### ISSUE TITLE
the loss is nan

### ISSUE THREAD

**Issue by zcdliuwei** [hunglc007/tensorflow-yolov4-tflite#14]
Hi,
when i run python train.py on coco data, follow your tutorial exactly:
i get :
=> STEP   49   lr: 0.000020   giou_loss: 9.46    conf_loss: 42.83   prob_loss: 6.38   total_loss: 58.67
=> STEP   50   lr: 0.000020   giou_loss:  nan     conf_loss:  nan     prob_loss:  nan   total_loss:  nan
=> STEP   51   lr: 0.000021    giou_loss:  nan      conf_loss:  nan     prob_loss:  nan   total_loss:  nan
=> STEP   52   lr: 0.000021    giou_loss:  nan     conf_loss:  nan     prob_loss:  nan   total_loss:  nan
All parameters I use are default。

in addition， I think the nine line of main function in train.py should be ：
steps_per_epoch = len(trainset) // cfg.TRAIN.BATCH_SIZE，

Your code is：
steps_per_epoch = len(trainset)

I am a beginner of computer vision, I don't know if my understanding is wrong ？

**Comment by jzoker** [comment]
We noticing the issue too, it is because of use of softplus without threshold. It'll updated soon.

**Comment by EdwinBaek** [comment]
Hi 
When I run train.py code on my custom data, I get nan value too.
Did you solve this problem now? I want to solution this problem, but I'm beginner of CV too... 

I think my dataset is a problem...

This is a example my train set & test set.  this is a text file.
img_path x_min,y_min,x_max,y_max,label_number
ex)
./image/train/Edwin.png 153,21,300,400,0     

is this alright to use dataset for training??

I need your help... Thanks you.

(Sorry for my English skill is worst.. I need you understand What I'm talking about..)

**Comment by bgamini** [comment]
I had same problem, solved it by changing learning rate to 1e-4 and using my own way for transfer learning [(here)](https://github.com/hunglc007/tensorflow-yolov4-tflite/issues/13#issuecomment-622875138)

**Comment by EdwinBaek** [comment]
> I had same problem, solved it by changing learning rate to 1e-4 and using my own way for transfer learning [(here)](https://github.com/hunglc007/tensorflow-yolov4-tflite/issues/13#issuecomment-622875138)

Thank you for your help! I solve this problem. Because of .weight file, I can't learning my own dataset. So I don't use yolov4.weight file. Just make a new .weights files. Thank you!

**Comment by zcdliuwei** [comment]
> I had same problem, solved it by changing learning rate to 1e-4 and using my own way for transfer learning [(here)](https://github.com/hunglc007/tensorflow-yolov4-tflite/issues/13#issuecomment-622875138)

which learning rate is set to 1e-4 ? lr_init or lr_end ? 
and even I convert weight files from .weights to .pkl, my loss is also non on coco dataset

**Comment by zcdliuwei** [comment]
> We're noticing the issue too, it is due to use of softplus without threshold. It'll be updated soon.

please let me know if you can do this, thanks

**Comment by axiaobugs** [comment]
I had same problem

**Comment by axiaobugs** [comment]
i found in yolov4.py use the sigmoid.

**Comment by lddsjy** [comment]
Same problem，too.

**Comment by chokyungjin** [comment]
me too, Same problems....

**Comment by ShihuaiXu** [comment]
me too, Same problems....

**Comment by zateyev** [comment]
I replaced the mish activation with leaky_relu in common.py and it worked for me

**Comment by ShihuaiXu** [comment]
thank you very much





------------------ Original ------------------

**Comment by zcdliuwei** [comment]
> I replaced the mish activation with leaky_relu in common.py and it worked for me

When training to more than one thousand iterations, the loss value is still Nan....

**Comment by zxj11838** [comment]
The 'bbox_iou' function on line 177 in 'yolov4.py' is wrong, This may be related to the loss of the network as nan.
https://github.com/hunglc007/tensorflow-yolov4-tflite/issues/24

**Comment by ShihuaiXu** [comment]
> I replaced the mish activation with leaky_relu in common.py and it worked for me

I use this method to train my yolov4, the loss is normal, but the ap is lower compared with using mish as the activation function.

**Comment by zcdliuwei** [comment]
> The 'bbox_iou' function on line 177 in 'yolov4.py' is wrong, This may be related to the loss of the network as nan.
> #24

so , How to modify ?

**Comment by zxj11838** [comment]
@zcdliuwei 
You need to modify：
left_up = tf.maximum(boxes1_coor[..., :2], boxes1_coor[..., :2]) 
to：
left_up = tf.maximum(boxes1_coor[..., :2], boxes2_coor[..., :2])

**Comment by ShihuaiXu** [comment]
@zxj11838  are you serious? how about  right_down = tf.minimum(boxes2_coor[..., 2:], boxes2_coor[..., 2:])

**Comment by zxj11838** [comment]
According to my understanding, here is the calculation of the coordinates of the upper left corner and lower right corner of the intersection of the two boxes.

**Comment by zxj11838** [comment]
I modified the following content according to the readme，the loss of training and testing is not nan.

Traning your own model
# Prepare your dataset
# If you want to train from scratch:
In config.py set FISRT_STAGE_EPOCHS=0 
# Run script:
python train.py

**Comment by zcdliuwei** [comment]
> I modified the following content according to the readme，the loss of training and testing is not nan.
> Prepare your dataset
> If you want to train from scratch:
> In config.py set FISRT_STAGE_EPOCHS=0

I think this warehouse is the best  implemented in Tensorflow2.X of yolov4 so far, but the problem that the loss value is Nan has not been solved,  I've tried to modify some things, such as training from scratch, modifying activation functions, etc，but they all didn't work,  now I don't know what to do anymore, because my loss value has always been Nan, can you send me a copy of your code ?

**Comment by remic33** [comment]
Same problem here, even after using leaky_relu, loss will be nan after some steps. 
Can't be use for training.

**Comment by zcdliuwei** [comment]
I want to ask for your advice. I want to identify the bottles in the trash can. The real image is about this:

![20200611112928_B_Q](https://user-images.githubusercontent.com/36565076/85197457-a445a680-b313-11ea-9c4b-7ca36c25fd85.jpg)

![20200611112934_B_Q](https://user-images.githubusercontent.com/36565076/85197459-aad41e00-b313-11ea-95dc-d7ea0840dee5.jpg)

![20200611113139_B_Q](https://user-images.githubusercontent.com/36565076/85197466-af98d200-b313-11ea-854b-8611af959c3a.jpg)

The most complex part of this scene is the dense shielding and stacking of bottles. Our purpose is to tell the customer who threw the bottle what kind of bottle he threw in, and it's better to return the location information of the bottle at the bottom of the trash can. At present, I use Yolo series for testing, and the effect is always unsatisfactory.

I want to know does you have any suggestions to improve the detection and classification ability in this scenario? 

Much obliged for your assistance.

**Comment by Archana2016** [comment]
I have the same problem with the git repo downloaded and installed last week. I tried the suggestion by @bgamini , but that did not help, perhaps because I am training my own data from scratch. I am giving more details of my case, and would really appreciate your suggestions.

My environment:

Python 3.7 /  NVIDIA GeForce RTX 3070/ cudart64_101.dll/ cudnn64_7.dll/ 
Other libraries according to according to requirements-gpu.txt  

I am trying to train my custom data (with 3 claases) from scratch (without using yolov4.weights file). I prepared the dataset with the required annotation files, following the example of voc_annotation.  I also updated the contents and paths of classes.names,and  train.txt / test.txt  in ./data/classes and ./data/dataset folders respectively. 

In addition, 
I updated  __C.TRAIN.FISRT_STAGE_EPOCHS    = 0 in config, and 
updated FLAGS.weights as 
flags.DEFINE_string('weights', None, 'pretrained weights') 

=> TEST STEP   36   giou_loss:  nan   conf_loss:  nan   prob_loss:  nan   total_loss:  nan
<core.dataset.Dataset object at 0x000002388F00FB70> 5
=> STEP   36/ 150   lr: 0.000917   giou_loss:  nan   conf_loss:  nan   prob_loss:  nan   total_loss:  nan
=> STEP   37/ 150   lr: 0.000911   giou_loss:  nan   conf_loss:  nan   prob_loss:  nan   total_loss:  nan
=> STEP   38/ 150   lr: 0.000905   giou_loss:  nan   conf_loss:  nan   prob_loss:  nan   total_loss:  nan
=> STEP   39/ 150   lr: 0.000898   giou_loss:  nan   conf_loss:  nan   prob_loss:  nan   total_loss:  nan
=> STEP   40/ 150   lr: 0.000891   giou_loss:  nan   conf_loss:  nan   prob_loss:  nan   total_loss:  nan
